### DISCUSSION TITLE
Synchronized sub-drivers for 4th order S-curve

### DISCUSSION THREAD

**Discussion by Juanduino** [winder/Universal-G-Code-Sender#2353] [Help] [answered]
Hey !

@winder @breiler 

I have come across a S-curve algorithm for multi-axis machines, with the potential to enable SimpleFOC sub-nodes, with UGS. 

Practically speaking, how can we sync sub-drivers running SimpleFOC?

Is it possible to have the first command reach each motor-controller MCU at the exact same time? If so, then we can use the zero time as the synchronizing factor. Any other idea/approach is welcome. At this point I’m just investigating. 

I’m starting this new discussion, since the last discussion about closed loop virtual machine was getting a bit weird. 

I believe this S-curve approach can actually be a viable way to have SimpleFOC nodes work w. UGS?

https://github.com/Juanduino/S4_CurvePlanner/tree/main

![IMG_1073](https://github.com/winder/Universal-G-Code-Sender/assets/4369438/6b451322-64c9-46a0-8d65-842829016b1b)

Best regards 

JASEP

**Comment by breiler** [comment]
What controller are you planning to use for this?

**Reply by breiler** [reply]
To clarify the question, what controller firmware (GRBL, FluidNC, Smoothieware, grblHAL)?

**Reply by Juanduino** [reply]
The controller is running SimpleFOC (Closed loop, Field Oriented Control)

https://hackaday.io/project/190597-the-field-stack-nema2334-foc

As I mentioned in the other thread, my initial focus was getting OpenPNP working with SimpleFOC and that should be possible, since the motion in OpenPNP does not need strict time-syncing. Using the most basic Gcode settings for OpenPNP, using M400, achieving 4th order S-curve should be within range, doing all the trajectory planning on the MCU.

The planner algorithm also suggest a way to “scale” the moves, in order for the different axis to reach the target position in a synchronized manner. This will involve placing some of the algorithm/math in the host. But we still need some level of time-synchronization between the host and the sub-nodes. 

Let’s do a example:

The Gcode has a move to G1 X20.0000 Y30.0000. The algorithm then uses max feedrate/velocity, jerk and acceleration to calculate the trajectory for both axis, finds the slowest one and scales the feedrate, jerk and velocity for the other axis to match.

 But in order for the two separate controllers to initiate the move at the same time, either the commands should reach the nodes simultaneously or some level of time syncing should be introduced.

**Reply by breiler** [reply]
This sounds like a job for a realtime controller, how does this apply to UGS?

**Reply by Juanduino** [reply]
I suppose there is a potential for UGS to enable SimpleFOC nodes, if the syncing is possible.

**Reply by Juanduino** [reply]
I mean, if Klipper can do it, why not UGS with 4th order S-curve?

**Reply by breiler** [reply, marked as answer]
It might be possible to implement a virtual controller in UGS in a similar way as Klipper, that keeps track of the gcode state, planning, positions and inputs. But I still don't think that UGS is the platform for this. 


The whole Klipper ecosystem consists of many known parts which can be controlled. For instance, its host running on an dedicated single board computer with a known OS and dependencies. Each controller board is running their own firmware which is communicating with the host through a lightweight protocol. Etc.

UGS runs on Windows x86, Windows x64, Linux x64, Linux ARM, Linux aarch64, MacOSX x64, MacOSX aarch64. This is a lot of architectures to get this working on and to support.  


You might as well implement this as an self contained application. Any integration with this controller application can be done using a known protocol (such as the GRBL protocol over TCP) which would allow you to use practically any sender you'd like.

**Reply by Juanduino** [reply]
Just a side note: The controller has a CAN FD transceiver, but I haven’t even started exploring it. Maybe it is the way to go, as it appears to have some inbuilt syncing in the protocol? 

Edit: Ok, I’m quite sure it is the way forward. As I understand it, all nodes receive the CAN commands at the same time, which effectively means that no syncing on the host/nodes is necessary. We can just bundle the relevant move commands, with the expectation, that they will reach the MCUs simultaneously.
